Unexpected early exit. This happens when Promises returned by plugins cannot resolve. Unfinished hook action(s) on exit:

```
- building client + server bundles...
🎈 SUMMARY 解析中...
🎈 SUMMARY 解析完成...
x Build failed in 1m 42s
✖ building client + server bundles...
build error:
Unexpected early exit. This happens when Promises returned by plugins cannot resolve. Unfinished hook action(s) on exit:
(vitepress-plugin-pagefind) load "\u0000virtual:pagefind"
Error: Unexpected early exit. This happens when Promises returned by plugins cannot resolve. Unfinished hook action(s) on exit:
(vitepress-plugin-pagefind) load "\u0000virtual:pagefind"
    at process.handleBeforeExit (file:///home/runner/work/taophilosophy.github.io/taophilosophy.github.io/node_modules/.pnpm/rollup@4.[17](https://github.com/taophilosophy/taophilosophy.github.io/actions/runs/9171633579/job/25216376212#step:10:18).2/node_modules/rollup/dist/es/shared/node-entry.js:20115:28)
    at Object.onceWrapper (node:events:634:26)
    at process.emit (node:events:5[19](https://github.com/taophilosophy/taophilosophy.github.io/actions/runs/9171633579/job/25216376212#step:10:20):28)
 ELIFECYCLE  Command failed with exit code 1.
Error: Process completed with exit code 1.
```

请问应该如何做才能看到具体的错误信息？项目地址 <https://github.com/taophilosophy/taophilosophy.github.io>。奇怪的是本地按同样版本的软件配置就会无限卡住，内存 CPU 都没有变化，只有 git action 会抛出上述错误。

Git action 见 <https://github.com/taophilosophy/taophilosophy.github.io/blob/main/.github/workflows/deploy.yml>

--- a/packages/theme/CHANGELOG.md
+++ b/packages/theme/CHANGELOG.md
@@ -6,7 +6,7 @@
 
 - chore: fix readme spelling error.
 - Updated dependencies
-  - vitepress-plugin-pagefind@0.4.0
+  - vitepress-plugin-pagefind@0.4.1
 
 ## 0.3.6
 

--- a/packages/vitepress-plugin-pagefind/CHANGELOG.md
+++ b/packages/vitepress-plugin-pagefind/CHANGELOG.md
@@ -1,10 +1,17 @@
 # vitepress-plugin-pagefind
 
+## 0.4.1
+
+### Patch Changes
+
+- feat: add manual attr
+
 ## 0.4.0
 
 ### Minor Changes
 
 The build speed has been significantly improved.
+
 - style: search text title overflow
 - style: empty title space
 - chore: meta build inject page

--- a/packages/vitepress-plugin-pagefind/README-zh.md
+++ b/packages/vitepress-plugin-pagefind/README-zh.md
@@ -223,6 +223,55 @@ pagefindPlugin({
 })
 ```
 
+### 示例7: 自定义生成索引的位置
+
+*如果插件未能正常在 buildEnd 阶段，执行生成索引的指令，也可以按此种方式调整配置*
+
+① 修改运行脚本
+
+在 vitepress 构建完成后，添加索引生成的脚本
+
+CLI 参数见： https://pagefind.app/docs/config-options/
+
+```json
+{
+  "scripts": {
+    "docs:build": "vitepress build docs && npx pagefind --site docs/.vitepress/dist --exclude-selectors div.aside,a.header-anchor"
+  }
+}
+```
+
+② add head script
+```ts
+import { defineConfig } from 'vitepress'
+import { pagefindPlugin } from 'vitepress-plugin-pagefind'
+
+export default defineConfig({
+  head: [
+    // 添加脚本，按实际情况指定索引脚本的位置
+    [
+      'script',
+      {},
+      `import('/pagefind/pagefind.js')
+        .then((module) => {
+          window.__pagefind__ = module
+          module.init()
+        })
+        .catch(() => {
+          // console.log('not load /pagefind/pagefind.js')
+        })`
+    ]
+  ],
+  vite: {
+    plugins: [pagefindPlugin({
+      // 设置 manul 属性为 true
+      manual: true
+    })]
+  },
+  lastUpdated: true
+})
+```
+
 更多可配置的选项请查看下面的完整配置项
 
 ## 完整配置项如下
@@ -312,6 +361,12 @@ interface SearchConfig {
      * @default false
      */
     ignorePublish?: boolean
+    /**
+     * 手动控制索引生成指令和资源加载脚本
+     * @see README.md 示例7
+     * @default false
+     */
+    manual?: boolean
 }
 ```
 </details>

--- a/packages/vitepress-plugin-pagefind/README-zh.md
+++ b/packages/vitepress-plugin-pagefind/README-zh.md
@@ -227,6 +227,13 @@ pagefindPlugin({
 
 *如果插件未能正常在 buildEnd 阶段，执行生成索引的指令，也可以按此种方式调整配置*
 
+```js
+pagefindPlugin({
+  // 设置 manul 属性为 true
+  manual: true
+})
+```
+
 ① 修改运行脚本
 
 在 vitepress 构建完成后，添加索引生成的脚本

--- a/packages/vitepress-plugin-pagefind/README-zh.md
+++ b/packages/vitepress-plugin-pagefind/README-zh.md
@@ -14,7 +14,7 @@
 
 ## 如何使用
 
-①: 安装插件
+①: 安装插件和必要的依赖
 ```sh
 pnpm add vitepress-plugin-pagefind pagefind
 # or
@@ -49,7 +49,10 @@ export default defineConfig({
 })
 ```
 
-**(可选)** ③: 搜索优化
+<details>
+<summary>
+(可选) ③: 中文搜索优化
+</summary>
 
 如果你的文档主要内容是中文，推荐做以下设置，优化一下搜索
 
@@ -70,6 +73,9 @@ export default defineConfig({
 ```
 
 详见下面的示例4了解为什么要这样做
+
+</details>
+
 ## 高级用法
 
 ### 示例1：自定义搜素框文案
@@ -100,7 +106,7 @@ pagefindPlugin({
 })
 ```
 
-**建议**：不设置的前提下，插件会默认使用 `vitepress` 配置中设置的 `lang` 值
+**提示**：不设置的前提下，插件会默认使用 `vitepress` 配置中设置的 `lang` 值
 ```ts
 import { defineConfig } from 'vitepress'
 
@@ -142,12 +148,11 @@ pagefindPlugin({
 
 如果你有更好的实现，欢迎分享
 
-#### 4.2 搜索结果优化
-可以关闭内置的搜索结果优化，使用`filter`方法自定义过滤行为
+#### 4.2 搜索结果过滤
+使用`filter`方法自定义过滤行为
 
 ```js
 pagefindPlugin({
-  resultOptimization: false,
   filter(searchItem, idx, originArray) {
     console.log(searchItem)
     return !searchItem.route.includes('404')
@@ -286,6 +291,7 @@ interface SearchConfig {
     /**
      * 搜索结果优化
      * @default false
+     * @deprecated
      */
     resultOptimization?: boolean
     /**

--- a/packages/vitepress-plugin-pagefind/README.md
+++ b/packages/vitepress-plugin-pagefind/README.md
@@ -221,6 +221,54 @@ pagefindPlugin({
 })
 ```
 
+### Example 7: Custom indexing location
+
+*If the plugin cannot execute normally in the buildEnd stage, or custom indexing file location.*
+
+① modify script
+
+add pagefind indexing generation script
+
+CLI Options See： https://pagefind.app/docs/config-options/
+
+```json
+{
+  "scripts": {
+    "docs:build": "vitepress build docs && npx pagefind --site docs/.vitepress/dist --exclude-selectors div.aside,a.header-anchor"
+  }
+}
+```
+
+② add head script
+```ts
+import { defineConfig } from 'vitepress'
+import { pagefindPlugin } from 'vitepress-plugin-pagefind'
+
+export default defineConfig({
+  head: [
+    // add script，manually specify the location of the import index file
+    [
+      'script',
+      {},
+      `import('/pagefind/pagefind.js')
+        .then((module) => {
+          window.__pagefind__ = module
+          module.init()
+        })
+        .catch(() => {
+          // console.log('not load /pagefind/pagefind.js')
+        })`
+    ]
+  ],
+  vite: {
+    plugins: [pagefindPlugin({
+      manual: true
+    })]
+  },
+  lastUpdated: true
+})
+```
+
 See options below for more details
 ## Options
 TS DTS see [src/type.ts](./src/type.ts)
@@ -308,6 +356,12 @@ interface SearchConfig {
      * @default false
      */
     ignorePublish?: boolean
+    /**
+     * Manually control index generation instructions and resource loading scripts
+     * @see README.md Example7
+     * @default false
+     */
+    manual?: boolean
 }
 ```
 </details>

--- a/packages/vitepress-plugin-pagefind/README.md
+++ b/packages/vitepress-plugin-pagefind/README.md
@@ -14,7 +14,7 @@ Offline full-text search based on [pagefind](https://github.com/cloudcannon/page
 
 ## Usage
 
-step1: install plugin
+step1: Install plugin and dependencies
 ```sh
 pnpm add vitepress-plugin-pagefind pagefind
 # or
@@ -49,7 +49,8 @@ export default defineConfig({
 })
 ```
 
-**(optional)** step3: customSearchQuery
+<details>
+<summary>step3: Chinese search optimization</summary>
 
 if your docs language（`lang`） is Chinese (`zh-`)
 
@@ -69,6 +70,8 @@ export default defineConfig({
 })
 ```
 see Example4 below for details
+</details>
+
 
 ## Advanced Usage
 
@@ -144,13 +147,11 @@ pagefindPlugin({
 
 If you have a better implementation, welcome to share
 
-#### 4.2 Search result optimization
-You can turn off the built-in search results optimization
+#### 4.2 Search result filter
+Use the `filter` method to customize the filtering behavior.
 
-Implement it yourself using the `filter` method
 ```js
 pagefindPlugin({
-  resultOptimization: false,
   filter(searchItem, idx, originArray) {
     console.log(searchItem)
     return !searchItem.route.includes('404')
@@ -286,6 +287,7 @@ interface SearchConfig {
     customSearchQuery?: (input: string) => string
     /**
      * @default false
+     * @deprecated
      */
     resultOptimization?: boolean
     /**

--- a/packages/vitepress-plugin-pagefind/README.md
+++ b/packages/vitepress-plugin-pagefind/README.md
@@ -225,6 +225,12 @@ pagefindPlugin({
 
 *If the plugin cannot execute normally in the buildEnd stage, or custom indexing file location.*
 
+```js
+pagefindPlugin({
+  manual: true
+})
+```
+
 ① modify script
 
 add pagefind indexing generation script

--- a/packages/vitepress-plugin-pagefind/package.json
+++ b/packages/vitepress-plugin-pagefind/package.json
@@ -1,6 +1,6 @@
 {
   "name": "vitepress-plugin-pagefind",
-  "version": "0.4.0",
+  "version": "0.4.1",
   "description": "vitepress plugin, Offline full-text search based on pagefind implementation.",
   "author": "sugar",
   "license": "MIT",

--- a/packages/vitepress-plugin-pagefind/src/index.ts
+++ b/packages/vitepress-plugin-pagefind/src/index.ts
@@ -57,6 +57,8 @@ export function pagefindPlugin(
         // 调用自己的
         await selfBuildEnd?.(siteConfig)
         await buildEnd(searchConfig)
+        const okMark = '\x1B[32m✓\x1B[0m'
+        console.log(`${okMark} generating pagefind Indexing...`)
       }
 
       // 通过 head 添加额外的脚本注入

--- a/packages/vitepress-plugin-pagefind/src/index.ts
+++ b/packages/vitepress-plugin-pagefind/src/index.ts
@@ -2,10 +2,10 @@ import path from 'node:path'
 import { fileURLToPath } from 'node:url'
 import { Buffer } from 'node:buffer'
 import type { PluginOption } from 'vite'
-import type { SiteConfig } from 'vitepress'
+import type { HeadConfig, SiteConfig } from 'vitepress'
 import { stringify } from 'javascript-stringify'
 import matter from 'gray-matter'
-import { getFileBirthTime, pluginSiteConfig } from './node'
+import { buildEnd, getFileBirthTime, getPagefindHead } from './node'
 import type { PagefindOption, SearchConfig } from './type'
 
 function isESM() {
@@ -51,24 +51,19 @@ export function pagefindPlugin(
         return
       }
 
-      // 添加 自定义 vitepress 的钩子
-
+      // 添加生成索引的方法
       const selfBuildEnd = vitepressConfig.buildEnd
-      vitepressConfig.buildEnd = (siteConfig: any) => {
+      vitepressConfig.buildEnd = async (siteConfig: any) => {
         // 调用自己的
-        selfBuildEnd?.(siteConfig)
-        siteConfig = Object.assign(siteConfig || {}, {
-          PagefindOption: searchConfig
-        })
-        pluginSiteConfig?.buildEnd?.(siteConfig)
+        await selfBuildEnd?.(siteConfig)
+        await buildEnd(searchConfig)
       }
 
+      // 通过 head 添加额外的脚本注入
       const selfTransformHead = vitepressConfig.transformHead
       vitepressConfig.transformHead = async (ctx) => {
         const selfHead = (await Promise.resolve(selfTransformHead?.(ctx))) || []
-        const pluginHead
-          = (await Promise.resolve(pluginSiteConfig?.transformHead?.(ctx))) || []
-        return selfHead.concat(pluginHead)
+        return selfHead.concat(getPagefindHead(ctx.siteData.base) as HeadConfig[])
       }
     },
     resolveId(id: string) {
@@ -80,22 +75,6 @@ export function pagefindPlugin(
     load(this, id) {
       if (id !== resolvedVirtualModuleId)
         return
-      // TODO: 历史逻辑，先保留注释
-      // const srcDir
-      //   = resolveConfig.vitepress.srcDir
-      //     .replace(resolveConfig.vitepress.root, '')
-      //     .replace(/^\//, '')
-      //   || process.argv.slice(2)?.[1]
-      //   || '.'
-
-      // let docsData: any[] = []
-      // if (runCommand === 'build') {
-      //   docsData = await getPagesData(
-      //     srcDir,
-      //     resolveConfig.vitepress,
-      //     searchConfig
-      //   )
-      // }
 
       return `
       import { ref } from 'vue'

--- a/packages/vitepress-plugin-pagefind/src/index.ts
+++ b/packages/vitepress-plugin-pagefind/src/index.ts
@@ -41,6 +41,9 @@ export function pagefindPlugin(
       }
     },
     configResolved(config: any) {
+      if (searchConfig.manual) {
+        return
+      }
       if (resolveConfig) {
         return
       }
@@ -89,7 +92,7 @@ export function pagefindPlugin(
       if (id.endsWith('theme-default/Layout.vue')) {
         return code.replace('<VPContent>', '<VPContent data-pagefind-body>')
       }
-      if (id.endsWith('.md')) {
+      if (id.endsWith('.md') && !searchConfig.manual) {
         const { data: frontmatter } = matter(code, {
           excerpt: true
         })

--- a/packages/vitepress-plugin-pagefind/src/node.ts
+++ b/packages/vitepress-plugin-pagefind/src/node.ts
@@ -136,60 +136,52 @@ export function withBase(base: string, path: string) {
     : joinPath(base, path)
 }
 
-export const pluginSiteConfig: Partial<SiteConfig> = {
-  /**
-   * TODO：支持更多pagefind配置项
-   * vitepress buildEnd钩子调用
-   */
-  buildEnd(ctx) {
-    const pagefindOps: PagefindOption = (ctx as any).PagefindOption
-    const ignore = [
-      ...new Set(ignoreSelectors.concat(pagefindOps?.excludeSelector || []))
-    ]
-    const { log } = console
-    log()
-    log('=== pagefind: https://pagefind.app/ ===')
-    const siteDir = path.join(
-      process.argv.slice(2)?.[1] || '.',
-      '.vitepress/dist'
-    )
-    let command = `npx pagefind --site ${siteDir}`
+export async function buildEnd(pagefindOps: PagefindOption) {
+  const ignore = [
+    ...new Set(ignoreSelectors.concat(pagefindOps?.excludeSelector || []))
+  ]
+  const { log } = console
+  log()
+  log('=== pagefind: https://pagefind.app/ ===')
+  const siteDir = path.join(
+    process.argv.slice(2)?.[1] || '.',
+    '.vitepress/dist'
+  )
+  let command = `npx pagefind --site ${siteDir}`
 
-    if (ignore.length) {
-      command += ` --exclude-selectors "${ignore.join(', ')}"`
-    }
-
-    if (typeof pagefindOps.forceLanguage === 'string') {
-      command += ` --force-language ${pagefindOps.forceLanguage}`
-    }
-    // 用户自定义指令
-    if (pagefindOps.indexingCommand) {
-      command = pagefindOps.indexingCommand
-    }
-    log(command)
-    log()
-    execSync(command, {
-      stdio: 'inherit'
-    })
-  },
-  transformHead(ctx) {
-    return [
-      [
-        'script',
-        {},
-        `import('${withBase(ctx.siteData.base || '', '/pagefind/pagefind.js')}')
-    .then((module) => {
-      window.__pagefind__ = module
-      module.init()
-    })
-    .catch(() => {
-      // console.log('not load /pagefind/pagefind.js')
-    })`
-      ]
-    ]
+  if (ignore.length) {
+    command += ` --exclude-selectors "${ignore.join(', ')}"`
   }
-}
 
+  if (typeof pagefindOps.forceLanguage === 'string') {
+    command += ` --force-language ${pagefindOps.forceLanguage}`
+  }
+  // 用户自定义指令
+  if (pagefindOps.indexingCommand) {
+    command = pagefindOps.indexingCommand
+  }
+  log(command)
+  log()
+  execSync(command, {
+    stdio: 'inherit'
+  })
+}
+export function getPagefindHead(base: string) {
+  return [
+    [
+      'script',
+      {},
+      `import('${withBase(base || '', '/pagefind/pagefind.js')}')
+  .then((module) => {
+    window.__pagefind__ = module
+    module.init()
+  })
+  .catch(() => {
+    // console.log('not load /pagefind/pagefind.js')
+  })`
+    ]
+  ]
+}
 export function chineseSearchOptimize(input: string) {
   return input
     .replace(/[\u4E00-\u9FA5]/g, ' $& ')

--- a/packages/vitepress-plugin-pagefind/src/search.ts
+++ b/packages/vitepress-plugin-pagefind/src/search.ts
@@ -46,9 +46,17 @@ interface Filters {
 }
 
 function decodeBase64AndDeserialize(base64String: string) {
-  const serialized = decodeURIComponent(atob(base64String))
-  const obj = JSON.parse(serialized)
-  return obj
+  if (!base64String) {
+    return {}
+  }
+  try {
+    const serialized = decodeURIComponent(atob(base64String))
+    const obj = JSON.parse(serialized)
+    return obj
+  }
+  catch {
+    return {}
+  }
 }
 
 export function formatPagefindResult(result: PagefindResult) {

--- a/packages/vitepress-plugin-pagefind/src/type.ts
+++ b/packages/vitepress-plugin-pagefind/src/type.ts
@@ -86,4 +86,11 @@ export interface SearchConfig {
    * @default false
    */
   ignorePublish?: boolean
+
+  /**
+   * Manually control index generation instructions and resource loading scripts
+   * @see README.md Example7
+   * @default false
+   */
+  manual?: boolean
 }

